tensorflow/core/kernels/segment_reduction_ops.cc

```diff
@@ -374,24 +373,104 @@ struct UnsortedSegmentFunctor<CPUDevice, T, Index, InitialValueF, ReductionF> {
                   typename TTypes<Index>::ConstFlat segment_ids,
                   const Index data_size, const T* data,
                   typename TTypes<T, 2>::Tensor output) {
-    output.setConstant(InitialValueF()());
+    auto cpu_device = ctx->eigen_cpu_device();
+    output.device(cpu_device) = output.constant(InitialValueF()());
     if (data_size == 0) {
       return;
     }
+
+    // This functor will reduce N rows input to M rows output
+    // N: segment_ids.dimension(0)
+    // M: num_segments
     const int64 N = segment_ids.dimension(0);
+    const int64 inner_dim = data_size / N;
     ReductionF reduction;
-    auto data_flat = typename TTypes<T, 2>::ConstTensor(data, N, data_size / N);
-    for (int64 i = 0; i < N; ++i) {
+    auto data_flat = typename TTypes<T, 2>::ConstTensor(data, N, inner_dim);
+    const int num_threads = cpu_device.numThreads();
+
+    // 'num_reductions' counts the number of output rows actually reduced,
+    // the rows only filled with InitialValueF() will be excluded.
+    // It also determines the degree of maximum parallelism.
+    int64 num_reductions = 0;
+    // 'row_counter' records how many input rows will be reduced in each
+    // output row, the row only fills with InitialValueF() will keep 0.
+    std::vector<Index> row_counter(num_segments, 0);
+    for (int64 i = 0; i < N; i++) {
+      // Get the corresponding output index j of input i.
       Index j = internal::SubtleMustCopy(segment_ids(i));
+      // Check the validity of index j.
       if (j < 0) {
         continue;
       }
       OP_REQUIRES(ctx, FastBoundsCheck(j, num_segments),
                   errors::InvalidArgument(
                       "segment_ids", SliceDebugString(segment_ids_shape, i),
                       " = ", j, " is out of range [0, ", num_segments, ")"));
-      reduction(data_flat.template chip<0>(i), output.template chip<0>(j));
+      if (row_counter[j] == 0) {
+        num_reductions++;
+      }
+      row_counter[j]++;
+    }
+    eigen_assert(num_reductions <= num_segments &&
+                 "reduced rows number can't be greater than num_segments.");
+
+    // Estimate task size and block number for shard function according to the
+    // rules:
+    // 1. each task size is 64 at least, to ensure each thread gets enough
+    //    work to do.
+    // 2. total task block number shouldn't be greater than num_reductions or
+    //    num_threads.
+    const int min_block_size = 64;
+    const int max_block_num = std::min(N / min_block_size + 1, num_reductions);
+    int block_num = std::min(max_block_num, num_threads);
+    const int block_size = N / block_num;
```

Comment:
could you please make it `int64` (and all other variables above), there are bunch of warnings from assigning int64->int

Reply:
Done.